File: endpoints/c/src/zerodds_wire.c
/* SPDX-License-Identifier: Apache-2.0
 * Copyright 2026 ZeroDDS Contributors
 *
 * zerodds_wire.c -- ISO C89 implementation of the endpoint wire-core.
 * See zerodds_wire.h. No malloc, no C99, byte-by-byte, host-endian-safe.
 */

#include "zerodds_wire.h"

/* ------------------------------------------------------------------ */
/* helpers                                                            */
/* ------------------------------------------------------------------ */

size_t zdw_padding_for(size_t pos, size_t align)
{
    size_t mask = align - 1u;
    return (align - (pos & mask)) & mask;
}

zdw_u64_t zdw_u64_from_ul(unsigned long v)
{
    zdw_u64_t r;
    r.hi = 0u;
    r.lo = v;
    return r;
}

/* Returns nonzero if the host stores multi-byte scalars big-endian first. */
static int host_is_be(void)
{
    unsigned long x = 1u;
    unsigned char *p = (unsigned char *)&x;
    return p[0] == 0; /* first byte is MSB on BE, LSB on LE */
}

/* Reverses n bytes in place. */
static void reverse_bytes(unsigned char *b, size_t n)
{
    size_t i = 0, j = (n == 0 ? 0 : n - 1);
    while (i < j) {
        unsigned char t = b[i];
        b[i] = b[j];
        b[j] = t;
        i++;
        j--;
    }
}

/* Normalizes host-memory bytes of a scalar to logical little-endian order. */
static void host_bytes_to_le(unsigned char *b, size_t n)
{
    if (host_is_be()) {
        reverse_bytes(b, n);
    }
}

/* ------------------------------------------------------------------ */
/* writer                                                             */
/* ------------------------------------------------------------------ */

void zdw_writer_init(zdw_writer *w, unsigned char *buf, size_t cap, int endian)
{
    w->buf = buf;
    w->cap = cap;
    w->len = 0u;
    w->endian = endian;
    w->max_align = ZDW_XCDR2_MAX_ALIGN;
    w->error = ZDW_OK;
}

static int w_align(zdw_writer *w, size_t align)
{
    size_t eff = align;
    size_t pad;
    if (eff > (size_t)w->max_align) {
        eff = (size_t)w->max_align;
    }
    pad = zdw_padding_for(w->len, eff);
    if (w->len + pad > w->cap) {
        w->error = ZDW_E_OVERrun;
        return ZDW_E_OVERrun;
    }
    while (pad-- > 0u) {
        w->buf[w->len++] = 0;
    }
    return ZDW_OK;
}

int zdw_put_bytes(zdw_writer *w, const unsigned char *data, size_t n)
{
    size_t i;
    if (w->error != ZDW_OK) {
        return w->error;
    }
    if (w->len + n > w->cap) {
        w->error = ZDW_E_OVERrun;
        return ZDW_E_OVERrun;
    }
    for (i = 0; i < n; i++) {
        w->buf[w->len++] = data[i];
    }
    return ZDW_OK;
}

/* Writes `n` logical-LE bytes at the given alignment, in wire byte order. */
static int w_put_uint(zdw_writer *w, size_t align, const unsigned char *le, size_t n)
{
    unsigned char tmp[8];
    size_t i;
    if (w->error != ZDW_OK) {
        return w->error;
    }
    if (w_align(w, align) != ZDW_OK) {
        return w->error;
    }
    for (i = 0; i < n; i++) {
        tmp[i] = le[i];
    }
    if (w->endian == ZDW_BE) {
        reverse_bytes(tmp, n);
    }
    return zdw_put_bytes(w, tmp, n);
}

int zdw_put_u8(zdw_writer *w, unsigned char v)
{
    if (w->error != ZDW_OK) {
        return w->error;
    }
    return zdw_put_bytes(w, &v, 1);
}

int zdw_put_u16(zdw_writer *w, unsigned int v)
{
    unsigned char le[2];
    le[0] = (unsigned char)(v & 0xffu);
    le[1] = (unsigned char)((v >> 8) & 0xffu);
    return w_put_uint(w, 2, le, 2);
}

int zdw_put_u32(zdw_writer *w, unsigned long v)
{
    unsigned char le[4];
    le[0] = (unsigned char)(v & 0xffu);
    le[1] = (unsigned char)((v >> 8) & 0xffu);
    le[2] = (unsigned char)((v >> 16) & 0xffu);
    le[3] = (unsigned char)((v >> 24) & 0xffu);
    return w_put_uint(w, 4, le, 4);
}

int zdw_put_u64(zdw_writer *w, zdw_u64_t v)
{
    unsigned char le[8];
    le[0] = (unsigned char)(v.lo & 0xffu);
    le[1] = (unsigned char)((v.lo >> 8) & 0xffu);
    le[2] = (unsigned char)((v.lo >> 16) & 0xffu);
    le[3] = (unsigned char)((v.lo >> 24) & 0xffu);
    le[4] = (unsigned char)(v.hi & 0xffu);
    le[5] = (unsigned char)((v.hi >> 8) & 0xffu);
    le[6] = (unsigned char)((v.hi >> 16) & 0xffu);
    le[7] = (unsigned char)((v.hi >> 24) & 0xffu);
    return w_put_uint(w, 8, le, 8);
}

int zdw_put_bool(zdw_writer *w, int v)
{
    return zdw_put_u8(w, (unsigned char)(v ? 1 : 0));
}

int zdw_put_f32(zdw_writer *w, float v)
{
    union { float f; unsigned char b[4]; } u;
    u.f = v;
    host_bytes_to_le(u.b, 4);
    return w_put_uint(w, 4, u.b, 4);
}

int zdw_put_f64(zdw_writer *w, double v)
{
    union { double d; unsigned char b[8]; } u;
    u.d = v;
    host_bytes_to_le(u.b, 8);
    return w_put_uint(w, 8, u.b, 8);
}

int zdw_put_string(zdw_writer *w, const char *s)
{
    size_t n = 0;
    unsigned char nul = 0;
    while (s[n] != '\0') {
        n++;
    }
    if (zdw_put_u32(w, (unsigned long)(n + 1u)) != ZDW_OK) {
        return w->error;
    }
    if (zdw_put_bytes(w, (const unsigned char *)s, n) != ZDW_OK) {
        return w->error;
    }
    return zdw_put_bytes(w, &nul, 1);
}

int zdw_put_seq_u8(zdw_writer *w, const unsigned char *data, size_t n)
{
    if (zdw_put_u32(w, (unsigned long)n) != ZDW_OK) {
        return w->error;
    }
    return zdw_put_bytes(w, data, n);
}

size_t zdw_dheader_begin(zdw_writer *w)
{
    unsigned char zero4[4];
    zero4[0] = 0; zero4[1] = 0; zero4[2] = 0; zero4[3] = 0;
    /* A DHEADER is a u32: 4-align, then reserve 4 bytes to back-patch. */
    if (w_align(w, 4) != ZDW_OK) {
        return w->len;
    }
    zdw_put_bytes(w, zero4, 4);
    return w->len; /* body starts here; the DHEADER is the 4 bytes before it */
}

int zdw_dheader_end(zdw_writer *w, size_t body_start)
{
    unsigned long len;
    unsigned char le[4];
    unsigned char *dst;
    if (w->error != ZDW_OK) {
        return w->error;
    }
    len = (unsigned long)(w->len - body_start);
    le[0] = (unsigned char)(len & 0xffu);
    le[1] = (unsigned char)((len >> 8) & 0xffu);
    le[2] = (unsigned char)((len >> 16) & 0xffu);
    le[3] = (unsigned char)((len >> 24) & 0xffu);
    if (w->endian == ZDW_BE) {
        reverse_bytes(le, 4);
    }
    dst = w->buf + (body_start - 4u);
    dst[0] = le[0]; dst[1] = le[1]; dst[2] = le[2]; dst[3] = le[3];
    return ZDW_OK;
}

int zdw_dheader_read(zdw_reader *r, unsigned long *len)
{
    return zdw_get_u32(r, len);
}

size_t zdw_emheader_begin(zdw_writer *w, unsigned long member_id,
                          int must_understand)
{
    /* EMHEADER = (M << 31) + (LC4 << 28) + member_id (28-bit). */
    unsigned long em = (must_understand ? 0x80000000uL : 0uL)
                     + (4uL << 28)
                     + (member_id & 0x0FFFFFFFuL);
    if (zdw_put_u32(w, em) != ZDW_OK) {
        return w->len;
    }
    /* LC4 NEXTINT = member body byte length: reuse the DHEADER back-patch. */
    return zdw_dheader_begin(w);
}

int zdw_emheader_end(zdw_writer *w, size_t body_start)
{
    return zdw_dheader_end(w, body_start);
}

int zdw_emheader_read(zdw_reader *r, unsigned long *member_id,
                      int *must_understand, unsigned long *nextint)
{
    unsigned long em = 0;
    if (zdw_get_u32(r, &em) != ZDW_OK) {
        return r->error;
    }
    *must_understand = ((em & 0x80000000uL) != 0uL);
    *member_id = em & 0x0FFFFFFFuL;
    return zdw_get_u32(r, nextint);
}

/* ------------------------------------------------------------------ */
/* reader                                                             */
/* ------------------------------------------------------------------ */

void zdw_reader_init(zdw_reader *r, const unsigned char *buf, size_t len, int endian)
{
    r->buf = buf;
    r->len = len;
    r->pos = 0u;
    r->endian = endian;
    r->max_align = ZDW_XCDR2_MAX_ALIGN;
    r->error = ZDW_OK;
}

static int r_align(zdw_reader *r, size_t align)
{
    size_t eff = align;
    size_t pad;
    if (eff > (size_t)r->max_align) {
        eff = (size_t)r->max_align;
    }
    pad = zdw_padding_for(r->pos, eff);
    if (r->pos + pad > r->len) {
        r->error = ZDW_E_OVERrun;
        return ZDW_E_OVERrun;
    }
    r->pos += pad;
    return ZDW_OK;
}

int zdw_get_bytes(zdw_reader *r, unsigned char *out, size_t n)
{
    size_t i;
    if (r->error != ZDW_OK) {
        return r->error;
    }
    if (r->pos + n > r->len) {
        r->error = ZDW_E_OVERrun;
        return ZDW_E_OVERrun;
    }
    for (i = 0; i < n; i++) {
        out[i] = r->buf[r->pos++];
    }
    return ZDW_OK;
}

/* Reads `n` bytes at the given alignment into logical-LE order. */
static int r_get_uint(zdw_reader *r, size_t align, unsigned char *le, size_t n)
{
    if (r->error != ZDW_OK) {
        return r->error;
    }
    if (r_align(r, align) != ZDW_OK) {
        return r->error;
    }
    if (zdw_get_bytes(r, le, n) != ZDW_OK) {
        return r->error;
    }
    if (r->endian == ZDW_BE) {
        reverse_bytes(le, n);
    }
    return ZDW_OK;
}

int zdw_get_u8(zdw_reader *r, unsigned char *out)
{
    return zdw_get_bytes(r, out, 1);
}

int zdw_get_u16(zdw_reader *r, unsigned int *out)
{
    unsigned char le[2];
    if (r_get_uint(r, 2, le, 2) != ZDW_OK) {
        return r->error;
    }
    *out = (unsigned int)le[0] | ((unsigned int)le[1] << 8);
    return ZDW_OK;
}

int zdw_get_u32(zdw_reader *r, unsigned long *out)
{
    unsigned char le[4];
    if (r_get_uint(r, 4, le, 4) != ZDW_OK) {
        return r->error;
    }
    *out = (unsigned long)le[0] | ((unsigned long)le[1] << 8) |
           ((unsigned long)le[2] << 16) | ((unsigned long)le[3] << 24);
    return ZDW_OK;
}

int zdw_get_u64(zdw_reader *r, zdw_u64_t *out)
{
    unsigned char le[8];
    if (r_get_uint(r, 8, le, 8) != ZDW_OK) {
        return r->error;
    }
    out->lo = (unsigned long)le[0] | ((unsigned long)le[1] << 8) |
              ((unsigned long)le[2] << 16) | ((unsigned long)le[3] << 24);
    out->hi = (unsigned long)le[4] | ((unsigned long)le[5] << 8) |
              ((unsigned long)le[6] << 16) | ((unsigned long)le[7] << 24);
    return ZDW_OK;
}

int zdw_get_bool(zdw_reader *r, int *out)
{
    unsigned char v;
    if (zdw_get_u8(r, &v) != ZDW_OK) {
        return r->error;
    }
    *out = (v != 0);
    return ZDW_OK;
}

int zdw_get_f32(zdw_reader *r, float *out)
{
    union { float f; unsigned char b[4]; } u;
    if (r_get_uint(r, 4, u.b, 4) != ZDW_OK) {
        return r->error;
    }
    /* u.b is logical-LE; put back into host memory order. */
    host_bytes_to_le(u.b, 4); /* symmetric: reverses again on BE host */
    *out = u.f;
    return ZDW_OK;
}

int zdw_get_f64(zdw_reader *r, double *out)
{
    union { double d; unsigned char b[8]; } u;
    if (r_get_uint(r, 8, u.b, 8) != ZDW_OK) {
        return r->error;
    }
    host_bytes_to_le(u.b, 8);
    *out = u.d;
    return ZDW_OK;
}

int zdw_get_string(zdw_reader *r, char *out, size_t cap)
{
    unsigned long len = 0u;
    size_t body;
    if (zdw_get_u32(r, &len) != ZDW_OK) {
        return r->error;
    }
    if (len == 0u) {
        r->error = ZDW_E_INVALID; /* length must include the NUL */
        return ZDW_E_INVALID;
    }
    if (len > (unsigned long)cap) {
        r->error = ZDW_E_OVERrun;
        return ZDW_E_OVERrun;
    }
    if (zdw_get_bytes(r, (unsigned char *)out, (size_t)len) != ZDW_OK) {
        return r->error;
    }
    body = (size_t)len - 1u;
    if (out[body] != '\0') {
        r->error = ZDW_E_INVALID; /* missing NUL terminator */
        return ZDW_E_INVALID;
    }
    return ZDW_OK;
}

int zdw_get_seq_u8(zdw_reader *r, unsigned char *out, size_t cap, size_t *n)
{
    unsigned long len = 0u;
    if (zdw_get_u32(r, &len) != ZDW_OK) {
        return r->error;
    }
    *n = (size_t)len;
    if ((size_t)len > cap) {
        r->error = ZDW_E_OVERrun;
        return ZDW_E_OVERrun;
    }
    return zdw_get_bytes(r, out, (size_t)len);
}

File: endpoints/python/zerodds_wire.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# zerodds_wire -- ZeroDDS endpoint wire-core, pure Python (ADR 0013).
#
# The `wire-core` unit of the native Python endpoint SDK: XCDR2 primitives
# byte-identical to the Rust core's zerodds-cdr. Pure stdlib, no C extension,
# conservative style (Python 2.7 / 3.x). Serialization is explicit-endian, so
# the output does not depend on the host byte order; the WIRE byte order is an
# explicit LE/BE parameter (the XCDR encapsulation flag, DDSI-RTPS 2.5 s10.5).
#
# XCDR2 alignment (OMG XTypes 1.3 s7.4.1.1.1): stream-relative, capped at 4.

from __future__ import print_function

import struct

LE = 0
BE = 1

_XCDR2_MAX_ALIGN = 4


def padding_for(pos, align):
    mask = align - 1
    return (align - (pos & mask)) & mask


class Writer(object):
    """Growing XCDR2 write buffer with an explicit wire byte order."""

    def __init__(self, endian):
        self.buf = bytearray()
        self.endian = endian
        self.max_align = _XCDR2_MAX_ALIGN

    def _align(self, align):
        eff = align if align <= self.max_align else self.max_align
        pad = padding_for(len(self.buf), eff)
        if pad:
            self.buf.extend(b"\x00" * pad)

    def put_bytes(self, data):
        self.buf.extend(data)

    def _put_uint(self, align, le_bytes):
        self._align(align)
        if self.endian == BE:
            self.buf.extend(le_bytes[::-1])
        else:
            self.buf.extend(le_bytes)

    def put_u8(self, v):
        self.buf.append(v & 0xFF)

    def put_u16(self, v):
        self._put_uint(2, struct.pack("<H", v & 0xFFFF))

    def put_u32(self, v):
        self._put_uint(4, struct.pack("<I", v & 0xFFFFFFFF))

    def put_u64(self, v):
        self._put_uint(8, struct.pack("<Q", v & 0xFFFFFFFFFFFFFFFF))

    def put_bool(self, v):
        self.put_u8(1 if v else 0)

    def put_f32(self, v):
        # IEEE-754 bit pattern via the u32 path, like the Rust f32::to_bits.
        self._put_uint(4, struct.pack("<f", v))

    def put_f64(self, v):
        self._put_uint(8, struct.pack("<d", v))

    def put_string(self, s):
        raw = s.encode("utf-8") if isinstance(s, type(u"")) else bytes(s)
        self.put_u32(len(raw) + 1)
        self.put_bytes(raw)
        self.put_u8(0)

    def put_seq_u8(self, data):
        self.put_u32(len(data))
        self.put_bytes(bytes(bytearray(data)))

    # --- DHEADER (XCDR2 delimited): @appendable/@mutable struct + collection
    #     with a non-primitive element. u32 (4-aligned) = body byte length. ---
    def dheader_begin(self):
        self._align(4)
        self.buf.extend(b"\x00\x00\x00\x00")
        return len(self.buf)  # body starts here

    def dheader_end(self, body_start):
        length = len(self.buf) - body_start
        le = struct.pack("<I", length & 0xFFFFFFFF)
        if self.endian == BE:
            le = le[::-1]
        self.buf[body_start - 4:body_start] = bytearray(le)

    # --- EMHEADER (XCDR2 @mutable, LC4): u32 = (M<<31)+(LC4<<28)+member_id,
    #     followed by NEXTINT (u32 = body length) + body. ---
    def emheader_begin(self, member_id, must_understand):
        em = (0x80000000 if must_understand else 0) + (4 << 28) \
            + (member_id & 0x0FFFFFFF)
        self.put_u32(em)
        return self.dheader_begin()  # NEXTINT slot + body start

    def emheader_end(self, body_start):
        self.dheader_end(body_start)

    def bytes(self):
        return bytes(self.buf)


class Reader(object):
    """XCDR2 read cursor over a bytes buffer."""

    def __init__(self, data, endian):
        self.buf = bytearray(data)
        self.pos = 0
        self.endian = endian
        self.max_align = _XCDR2_MAX_ALIGN

    def _align(self, align):
        eff = align if align <= self.max_align else self.max_align
        pad = padding_for(self.pos, eff)
        if self.pos + pad > len(self.buf):
            raise ValueError("overrun on align")
        self.pos += pad

    def get_bytes(self, n):
        if self.pos + n > len(self.buf):
            raise ValueError("overrun")
        out = bytes(self.buf[self.pos:self.pos + n])
        self.pos += n
        return out

    def _get_uint(self, align, n):
        self._align(align)
        raw = self.get_bytes(n)
        if self.endian == BE:
            raw = raw[::-1]
        return raw

    def get_u8(self):
        return bytearray(self.get_bytes(1))[0]

    def get_u16(self):
        return struct.unpack("<H", self._get_uint(2, 2))[0]

    def get_u32(self):
        return struct.unpack("<I", self._get_uint(4, 4))[0]

    def get_u64(self):
        return struct.unpack("<Q", self._get_uint(8, 8))[0]

    def get_bool(self):
        return self.get_u8() != 0

    def get_f32(self):
        return struct.unpack("<f", self._get_uint(4, 4))[0]

    def get_f64(self):
        return struct.unpack("<d", self._get_uint(8, 8))[0]

    def get_string(self):
        n = self.get_u32()
        if n == 0:
            raise ValueError("string length must include the NUL")
        raw = self.get_bytes(n)
        if bytearray(raw)[n - 1] != 0:
            raise ValueError("missing NUL terminator")
        return raw[:n - 1].decode("utf-8")

    def get_seq_u8(self):
        n = self.get_u32()
        return bytearray(self.get_bytes(n))

    def dheader_read(self):
        return self.get_u32()

    def emheader_read(self):
        em = self.get_u32()
        member_id = em & 0x0FFFFFFF
        must_understand = (em & 0x80000000) != 0
        nextint = self.get_u32()
        return (member_id, must_understand, nextint)
